common/service/service.go

```diff
@@ -58,24 +60,25 @@ type (
 		Logger          log.Logger
 		ThrottledLogger log.Logger
 
-		MetricScope         tally.Scope
-		MembershipFactory   MembershipMonitorFactory
-		RPCFactory          common.RPCFactory
-		PProfInitializer    common.PProfInitializer
-		PersistenceConfig   config.Persistence
-		ClusterMetadata     cluster.Metadata
-		ReplicatorConfig    config.Replicator
-		MetricsClient       metrics.Client
-		MessagingClient     messaging.Client
-		ESClient            es.Client
-		ESConfig            *es.Config
-		DynamicConfig       dynamicconfig.Client
-		DispatcherProvider  client.DispatcherProvider
-		DCRedirectionPolicy config.DCRedirectionPolicy
-		PublicClient        workflowserviceclient.Interface
-		ArchivalMetadata    archiver.ArchivalMetadata
-		ArchiverProvider    provider.ArchiverProvider
-		Authorizer          authorization.Authorizer
+		MetricScope              tally.Scope
+		MembershipFactory        MembershipMonitorFactory
+		RPCFactory               common.RPCFactory
+		AbstractDatastoreFactory persistenceClient.AbstractDataStoreFactory
```

Comment:
Summarizing what we discussed offline, there are two possible approaches here:

(1) Allow passing AbstractFactory here i.e. what this PR currently does
(2) Allow passing an override for the persistence factory as a whole. With this approach, the internal repo will have its own implementation of persistence factory - which knows how to build internal data store as well as other data stores that persistence.factoryImpl() knows how to build i.e. it will use composition 

The disadvantage of (1) is that you have to pass a map[string]string connectAttributes in CustomDataStoreConfig. 
The disadvantage of (2) is that there is some additional code to re-implement factory interface within the internal repo.

After looking at this PR, I am getting convinced that the amount of additional code is going to be smallish if we go with (2) i.e. both (1) and (2) are simple enough to implement. Given this, I prefer (2) because it looks cleaner in that the public repo won't have these customXXX / opaque maps  that are basically pass through. 

Let me know what you think.

Reply:
nvm - your current approach can also be made to work without having to pass map[string]string connectAttributes. i.e. both approaches will need to pass CustomDataStoreConfig - I don't see a way around it. So, your current approach looks just fine to me.

Reply:
That's right.